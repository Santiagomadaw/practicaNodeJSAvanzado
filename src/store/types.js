export const AUTH_LOGIN_PENDING = 'auth/login/pending';
export const AUTH_LOGIN_FULFILLED = 'auth/login/fulfilled';
export const AUTH_LOGIN_REJECTED = 'auth/login/rejected';

export const AUTH_LOGOUT_PENDING = 'auth/logout/pending';
export const AUTH_LOGOUT_FULFILLED = 'auth/logout/fulfilled';
export const AUTH_LOGOUT_REJECTED = 'auth/logout/rejected';


export const ADS_LOADED_PENDING = 'ads/loaded/pending';
export const ADS_LOADED_FULFILLED = 'ads/loaded/fulfilled';
export const ADS_LOADED_REJECTED = 'ads/loaded/rejected';

export const ADS_DELETED_PENDING = 'ads/deleted/pending';
export const ADS_DELETED_FULFILLED = 'ads/deleted/fulfilled';
export const ADS_DELETED_REJECTED = 'ads/deleted/rejected';

export const AD_DETAIL_PENDING = 'ad/detail/pending';
export const AD_DETAIL_FULFILLED = 'ad/detail/fulfilled';
export const AD_DETAIL_REJECTED = 'ad/detail/rejected';

export const AD_CREATED_PENDING = 'ad/created/pending';
export const AD_CREATED_FULFILLED = 'ad/created/fulfilled';
export const AD_CREATED_REJECTED = 'ad/created/rejected';

export const UPDATE_FILTER_SEARCH = 'update/filters/search';
export const UPDATE_FILTER_TAGS = 'update/filters/tags';
export const UPDATE_FILTER_BUYSELL = 'update/filters/buysell';
export const UPDATE_FILTER_PRICE = 'update/filters/price';
export const CLEAR_FILTERS = 'clear/filters';

export const UPDATE_SLIDER = 'update/slider';

export const UPDATE_TAGS_PENDING = 'update/tags/pending';
export const UPDATE_TAGS_FULFILLED = 'update/tags/fulfilled';
export const UPDATE_TAGS_REJECTED = 'update/tags/rejected';

export const UI_RESET_ERROR = 'ui/reset-error';
export const UI_SET_ERROR = 'ui/set-error';