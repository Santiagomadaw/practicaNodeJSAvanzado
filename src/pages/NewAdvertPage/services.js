import { client } from '../../utils/api/client';

export const postAd = async (formvalues) => {
    try {
        return await client.post('/api/v1/adverts', formvalues, {
            headers: { 'Content-Type': 'multipart/form-data' },
        });
    } catch (error) {
        
        return Promise.reject( error );
    }
};

