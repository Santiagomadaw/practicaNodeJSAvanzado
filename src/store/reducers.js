import {
  ADS_DELETED_FULFILLED,
  ADS_LOADED_FULFILLED,
  AD_CREATED_FULFILLED,
  AD_DETAIL_FULFILLED,
  AUTH_LOGIN_FULFILLED,
  AUTH_LOGOUT,
  CLEAR_FILTERS,
  UI_RESET_ERROR,
  UPDATE_FILTER_BUYSELL,
  UPDATE_FILTER_PRICE,
  UPDATE_FILTER_SEARCH,
  UPDATE_FILTER_TAGS,
  UPDATE_SLIDER,
  UPDATE_TAGS_FULFILLED,
} from './types';


const defaultState = {
  auth: false,
  ads: {
    loaded: false,
    data: [],
  },
  filters: {
    search: '',
    tags: [],
    buysell: 'all',
    price: [],
  },
  slider:
    [],
  ui: {
    pending: false,
    error: null,
  },
  tags: {
    data: [],
    loaded: false
  }
};
//----------------Login reducer--------------

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_FULFILLED:
      return true;
    case AUTH_LOGOUT:
      return false;
    default:
      return state;
  }
}
//----------------Ads reducer--------------

export function ads(state = defaultState.ads, action) {
  switch (action.type) {
    case ADS_LOADED_FULFILLED:
      return { loaded: true, data: action.payload };
    case AD_DETAIL_FULFILLED:
      return { loaded: false, data: [action.payload, ...state.data] };
    case AD_CREATED_FULFILLED:
      return { ...state, data: [action.payload, ...state.data] };
    case ADS_DELETED_FULFILLED:
      return { ...state, data: state.data.filter(ad => ad.id !== action.payload) };
    default:
      return state;
  }
}




//----------------Filters reducer--------------


export function filters(state = defaultState.filters, action) {
  switch (action.type) {
    case UPDATE_FILTER_SEARCH:

      return { ...state, search: action.payload };
    case UPDATE_FILTER_TAGS:
      return { ...state, tags: action.payload };
    case UPDATE_FILTER_BUYSELL:
      return { ...state, buysell: action.payload };
    case UPDATE_FILTER_PRICE:
      return { ...state, price: action.payload };
    case CLEAR_FILTERS:
      return {
        search: '',
        tags: [],
        buysell: 'all',
        price: action.payload
      };

    default:
      return state;
  }
}
//----------------Slider reducer--------------

export function slider(state = defaultState.slider, action) {
  switch (action.type) {
    case UPDATE_SLIDER:
      return [0, action.payload];
    default:
      return state;
  }
}
//----------------TAGS reducer--------------

export function tags(state = defaultState.tags, action) {
  switch (action.type) {
    case UPDATE_TAGS_FULFILLED:
      return { loaded: true, data: action.payload };
    default:
      return state;
  }
}




//----------------UI reducer--------------

export function ui(state = defaultState.ui, action) {
  if (action.error) {
    return { ...state, pending: false, error: action.payload };
  }
  if (action.type === UI_RESET_ERROR) {
    console.log(action.type)
    return { ...state, error: null };
  }
  if (action.type.endsWith('/pending')) {
    return { ...state, pending: true };
  }
  if (action.type.endsWith('/fulfilled')) {
    return { ...state, pending: false, error: null };
  }
  return state;
}
