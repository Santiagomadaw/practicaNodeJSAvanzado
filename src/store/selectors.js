export const getIsLogged = state => state.auth;


export const getLoadedAds = state => state.ads.data;

export const isLoadedAds = state => state.ads.loaded;

export const getAdByID = adId => state => getLoadedAds(state).find(ad => ad.id === adId);



export const getFilters = state => state.filters;

export const getSlider = state => state.slider;



export const getError = state => state.ui.error;

export const getPending = state => state.ui.pending;


export const getTagsLoaded = state => state.tags.data;

export const isLoadedTags = state => state.tags.loaded;
