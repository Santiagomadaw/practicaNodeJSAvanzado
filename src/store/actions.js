import {
    AUTH_LOGIN_FULFILLED,
    AUTH_LOGIN_REJECTED,
    AUTH_LOGIN_PENDING,
    ADS_LOADED_PENDING,
    ADS_LOADED_FULFILLED,
    ADS_LOADED_REJECTED,
    UPDATE_FILTER_SEARCH,
    UPDATE_FILTER_TAGS,
    UPDATE_FILTER_BUYSELL,
    UPDATE_FILTER_PRICE,
    CLEAR_FILTERS,
    UPDATE_SLIDER,
    UI_RESET_ERROR,
    AD_CREATED_PENDING,
    AD_CREATED_FULFILLED,
    AD_CREATED_REJECTED,
    AD_DETAIL_PENDING,
    AD_DETAIL_FULFILLED,
    AD_DETAIL_REJECTED,
    UPDATE_TAGS_PENDING,
    UPDATE_TAGS_FULFILLED,
    UPDATE_TAGS_REJECTED,
    ADS_DELETED_PENDING,
    ADS_DELETED_FULFILLED,
    ADS_DELETED_REJECTED,
    UI_SET_ERROR,
    AUTH_LOGOUT_PENDING,
    AUTH_LOGOUT_FULFILLED,
    AUTH_LOGOUT_REJECTED,
} from './types';
import { getAdByID, getLoadedAds, isLoadedAds, isLoadedTags } from './selectors';

//----------------Login actions--------------

export const authLogin = (credentials) => async (dispatch, _getState, { services: { login }, router }) => {
    try {
        dispatch(loginPending());
        await login(credentials);
        dispatch(loginFulfilled());
        const to = router.state.location.state?.from || '/';
        router.navigate(to, { replace: true });

    } catch (error) {
        if (error) {
            dispatch(loginRejected(error));
        }
    }
};

export const loginPending = () => ({
    type: AUTH_LOGIN_PENDING,
});
export const loginFulfilled = () => ({
    type: AUTH_LOGIN_FULFILLED,
});
export const loginRejected = (error) => ({
    type: AUTH_LOGIN_REJECTED,
    payload: error,
    error: true,
});

//----------------Logout actions--------------


export const logoutPending = () => ({
    type: AUTH_LOGOUT_PENDING,
});
export const logoutFulfilled = () => ({
    type: AUTH_LOGOUT_FULFILLED,
});
export const logoutRejected = (error) => ({
    type: AUTH_LOGOUT_REJECTED,
    payload: error,
    error: true,
});
export const authLogout = () => async (dispatch, _getState, { services: { logout } }) => {
    try {
        dispatch(logoutPending());
        await logout();
        dispatch(logoutFulfilled());

    } catch (error) {
        console.log(error);
        dispatch(logoutRejected(error));

    }
};

//----------------Ads load actions---------------

export const adsLoader = () => async (dispatch, getState, { services: { getAds } }) => {
    const haveAdsLoaded = isLoadedAds(getState());
    if (!haveAdsLoaded) {

        try {
            dispatch(adsLoadedPending());
            const rawAds = await getAds();
            dispatch(adsLoadedFulfilled(rawAds.data));
            dispatch(setSlider());
        } catch (error) {
            if (error) {
                console.log(error);
                dispatch(adsLoadedRejected(error));
            }
        }
    }
    dispatch(setSlider());
};
export const adsLoadedPending = () => ({
    type: ADS_LOADED_PENDING,
});

export const adsLoadedFulfilled = (ads) => ({
    type: ADS_LOADED_FULFILLED,
    payload: ads,
});

export const adsLoadedRejected = (error) => ({
    type: ADS_LOADED_REJECTED,
    payload: error,
    error: true,
});
//----------------Ad detail actions---------------

export const adDetailLoader = (id) => async (dispatch, getState, { services: { getAd } }) => {
    const state = getState();
    if (!getAdByID(id)(state)) {
        try {
            dispatch(adDetailPending());
            const ad = await getAd(id);
            dispatch(adDetailFulfilled(ad.data));
        } catch (error) {
            dispatch(adDetailRejected(error));
        }
    }

};
export const adDetailPending = () => ({
    type: AD_DETAIL_PENDING,
});

export const adDetailFulfilled = (ads) => ({
    type: AD_DETAIL_FULFILLED,
    payload: ads,
});

export const adDetailRejected = (error) => ({
    type: AD_DETAIL_REJECTED,
    payload: error,
    error: true,
});
//----------------Ad create actions---------------

export const adCreate = (formValues) => async (dispatch, _getState, { services: { postAd }, router }) => {

    try {
        dispatch(adCreatedPending());
        const response = await postAd(formValues);
        dispatch(adCreatedFulfilled(response.data));
        dispatch(setSlider());
        setTimeout(() => {

            const to = `/adverts/${response.data.id}`;
            router.navigate(to, { replace: true });
        }, 1000);


    } catch (error) {
        if (error) {
            dispatch(adCreatedRejected(error));
        }
    }


};

export const adCreatedPending = () => ({
    type: AD_CREATED_PENDING
});

export const adCreatedFulfilled = (ad) => ({
    type: AD_CREATED_FULFILLED,
    payload: ad,
});

export const adCreatedRejected = (error) => ({
    type: AD_CREATED_REJECTED,
    payload: error,
    error: true,
});

//----------------Ad delete actions---------------

export const adDelete = (id) => async (dispatch, _getState, { services: { deleteAd }, router }) => {

    try {

        dispatch(adDeletePending());
        if (id) {
            await deleteAd(id);
            dispatch(uiSetError('Anuncio Borrado'));
            setTimeout(() => {
                router.navigate('/adverts');
                dispatch(uiResetError());
            }, 1000);

            setTimeout(() => {
                dispatch(adDeleteFulfilled(id));
                dispatch(setSlider());
            }, 1200);

        }
    } catch (error) {
        if (error) {
            const msg = error.message;
            dispatch(adDeleteRejected(msg));
        }
    }



};

export const adDeletePending = () => ({
    type: ADS_DELETED_PENDING
});

export const adDeleteFulfilled = (id) => ({
    type: ADS_DELETED_FULFILLED,
    payload: id
});

export const adDeleteRejected = (error) => ({
    type: ADS_DELETED_REJECTED,
    payload: error,
    error: true,
});

//----------------Filters actions---------------


export const updateFilterSearch = (search) => ({
    type: UPDATE_FILTER_SEARCH,
    payload: search,
});
export const updateFilterTags = (tags) => ({
    type: UPDATE_FILTER_TAGS,
    payload: tags,
});
export const updateFilterBuysell = (buySell) => ({
    type: UPDATE_FILTER_BUYSELL,
    payload: buySell,
});
export const updateFilterPrice = (price) => ({
    type: UPDATE_FILTER_PRICE,
    payload: price,
});

export const clearFilters = (maxPrice) => ({
    type: CLEAR_FILTERS,
    payload: maxPrice,
});

export const updateSlider = (maxPrice) => ({
    type: UPDATE_SLIDER,
    payload: maxPrice,
});

export const setSlider = () => (dispatch, getState, { findHighestPrice }) => {
    const state = getState();
    const ads = getLoadedAds(state);
    const maxPrice = findHighestPrice(ads);
    dispatch(updateSlider(maxPrice));
    dispatch(updateFilterPrice([0, maxPrice]));
};
//----------------tags actions---------------

export const tagsLoader = () => async (dispatch, getState, { services: { getTags }, }) => {
    const haveTagsLoaded = isLoadedTags(getState());
    if (!haveTagsLoaded) {

        try {
            dispatch(tagsLoaderPending());
            const tags = await getTags();
            dispatch(tagsLoaderFulfilled(tags.data));


        } catch (error) {
            if (error) {
                dispatch(tagsLoaderRejected(error.message));
            }
        }
    }

};

export const tagsLoaderPending = () => ({
    type: UPDATE_TAGS_PENDING,
});
export const tagsLoaderFulfilled = (tags) => ({
    type: UPDATE_TAGS_FULFILLED,
    payload: tags,
});
export const tagsLoaderRejected = (error) => ({
    type: UPDATE_TAGS_REJECTED,
    payload: error,
    error: true,
});



export const uiResetError = () => ({
    type: UI_RESET_ERROR,
});
export const uiSetError = (error) => ({
    type: UI_SET_ERROR,
    payload: error,
    error: true,
});
