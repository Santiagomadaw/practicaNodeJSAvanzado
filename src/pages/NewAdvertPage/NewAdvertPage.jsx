import  { useEffect, useState } from 'react';
import Button from '../../components/shared/Button.jsx';
import Form from '../../components/shared/Form.jsx';
import Layout from '../../components/layout/Layout.jsx';
import FormField from '../../components/shared/FormField.jsx';
import './NewAdvertPage.css';
import RawSwitch from '../../components/shared/Switch.jsx';
import Select from '../../components/shared/Select.jsx';
import { useLocation, useNavigate } from 'react-router-dom';
import ErrorMessage from '../../components/shared/ErrorMessage.jsx';
import { useDispatch, useSelector } from 'react-redux';
import { adCreate, tagsLoader } from '../../store/actions.js';
import { getTagsLoaded } from '../../store/selectors.js';

export default function NewAdvertPage() {
    const location = useLocation();
    const dispatch = useDispatch()
    const navigate = useNavigate();
    const optionTags = useSelector(getTagsLoaded)
    const [formValues, setFormValues] = useState({
        name: '',
        price: 0,
        tags: '',
        sale: false,
    });
    const [error, setError] = useState(null);
    const resetError = () => setError(null);

    const { name, price, tags, sale } = formValues;

    const buttonDisabled = !name || !price || !tags.length;
    const handleChange = (event) => {
        setFormValues((currentFormValues) => ({
            ...currentFormValues,
            [event.target.name]: event.target.value,
        }));
    };
    const handleChangeMultiSelect = (event) => {
        const options = Array.from(event.target.selectedOptions).map(
            (option) => option.value,
        );
        setFormValues((currentFormValues) => ({
            ...currentFormValues,
            [event.target.name]: options.join(','),
        }));
    };
    const handleSwitchChange = (event) => {
        setFormValues((currentFormValues) => ({
            ...currentFormValues,
            [event.target.name]: event.target.checked,
        }));
    };
    function hasValidExtension(fileName) {
        const validExtensions = ['jpg', 'jpeg', 'png', 'gif', 'bmp', 'svg'];
        const extension = fileName
            .slice(((fileName.lastIndexOf('.') - 1) >>> 0) + 2)
            .toLowerCase();
        return validExtensions.includes(extension);
    }
    const handleFileUpload = (event) => {
        const file = event.target.files && event.target.files[0];
        if (file && hasValidExtension(file.name)) {
            const formData = new FormData();
            formData.append('photo', file);
            setFormValues((currentFormValues) => ({
                ...currentFormValues,
                [event.target.name]: file,
            }));
        } else {
            setError('Formato de imagen no valido');
        }
    };
    const handleBack = () => {
        const to = location.state?.from || '/';
        navigate(to, { replace: true });
    };
    const handleSubmit = async (event) => {
        event.preventDefault();
       dispatch(adCreate(formValues)) 
    };
    useEffect(() => {
        const getDataTags = async () => {
            dispatch(tagsLoader())
        };
        getDataTags();
    }, [dispatch]);
    return (
        
            <section className='new-ad-contianer'>
                <Form
                    id='ad-form'
                    $variant='column'
                    $customwidth='350px'
                >
                    <label htmlFor='name'>
                        <h4>Articulo:</h4>
                    </label>
                    <FormField
                        type='text'
                        name='name'
                        id='name'
                        role='name'
                        $customwidth='100%'
                        onChange={handleChange}
                        required
                    />
                    <label htmlFor='price'>
                        <h4>Price:</h4>
                    </label>
                    <FormField
                        type='number'
                        step='0.01'
                        name='price'
                        role='price'
                        id='price'
                        $customwidth='100%'
                        onChange={handleChange}
                        required
                    />
                    <label htmlFor='tags'>
                        <h4>Etiquetas:</h4>
                    </label>
                    <Select
                        name='tags'
                        role='tags'
                        id='tags'
                        $customwidth='100%'
                        $customheight='inered'
                        multiple
                        onChange={handleChangeMultiSelect}
                        required
                    >
                        {optionTags.map((tag, index) => (
                            <option
                                value={tag}
                                id={tag}
                                key={index}
                            >
                                {tag}
                            </option>
                        ))}
                    </Select>
                    <label htmlFor='photo'>
                        <h4>Foto:</h4>
                    </label>
                    <input
                        type='file' 
                        name='photo'
                        id='photo'
                        onChange={handleFileUpload}
                        accept='image/png, image/jpeg'
                    />
                    <RawSwitch
                        Name='sale'

                        role='sale'
                        Leftname='Compra'
                        Rightname='Venta'
                        checked={sale}
                        onChange={handleSwitchChange}
                    />
                    <div className='buttonWrapper'>
                        <Button
                            type='button'
                            role='buttonBack'
                            className='backButton'
                            $customwidth='50%'
                            onClick={handleBack}
                        >
                            Volver
                        </Button>
                        <Button
                            type='submit'
                            role='buttonSubmit'
                            disabled={buttonDisabled}
                            className='newAdButton'
                            $customwidth='50%'
                            onClick={handleSubmit}
                        >
                            Crear anuncio
                        </Button>
                    </div>
                    {error && (
                        <ErrorMessage
                            className='loginPage-error'
                            onClick={resetError}
                        >
                            <h3>{error.toUpperCase()}</h3>
                        </ErrorMessage>
                    )}
                </Form>
            </section>
        
    );
}
