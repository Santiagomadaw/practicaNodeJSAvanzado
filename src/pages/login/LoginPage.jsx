import { useState } from 'react';
import Form from '../../components/shared/Form';
import Layout from '../../components/layout/Layout';
import FormField from '../../components/shared/FormField';
import RawSwitch from '../../components/shared/Switch';
import styled from 'styled-components';
import ErrorMessage from '../../components/shared/ErrorMessage';
import Button from '../../components/shared/Button';
import { useDispatch, useSelector } from 'react-redux';
import {
    authLogin,
    uiResetError,
} from '../../store/actions';
import { getError, getPending } from '../../store/selectors';
import { Loading } from '../../components/shared/Loading';

export default function LoginPage() {
    const dispatch = useDispatch();
    const resetError = () => dispatch(uiResetError());
    const errorState = useSelector(getError);
    const pending = useSelector(getPending);
    const [formValues, setFormValues] = useState({
        email: '',
        password: '',
        save: false,
    });

    const handleChange = (event) => {
        const { name, value } = event.target;
        resetError();
        setFormValues((currentFormValues) => ({
            ...currentFormValues,
            [name]: value,
        }));
    };
    const handleSwitchChange = (event) => {
        const { name, checked } = event.target;
        resetError();
        setFormValues((currentFormValues) => ({
            ...currentFormValues,
            [name]: checked,
        }));
    };

    const handleSubmit = async (event) => {
        event.preventDefault();
        dispatch(authLogin(formValues))
       
        
    };

    const { email, password } = formValues;
    const buttonDisabled = !email || !password;
    return (
        <Layout>
            <StyledLogin className='loginPage'>
                <h1 className='loginPage-title'>Login</h1>
                <Form id='login-form' $variant='column' $customwidth='100%'>
                    <FormField
                        $customheight='38px'
                        $customwidth='100%'
                        type='text'
                        name='email'
                        id='email'
                        onChange={handleChange}
                    />
                    <FormField
                        $customheight='38px'
                        $customwidth='100%'
                        type='password'
                        name='password'
                        id='password'
                        onChange={handleChange}
                    />
                    <RawSwitch
                        Name='save'
                        checked={true}
                        onChange={handleSwitchChange}
                        Rightname='Guardar'
                    />
                    <Button
                        type='submit'
                        disabled={buttonDisabled}
                        className='loginButton'
                        onClick={handleSubmit}
                        $customwidth='100%'
                    >
                        Login
                    </Button>
                </Form>
                {pending && <Loading />}
            {errorState && (
                <ErrorMessage className='loginPage-error' onClick={resetError}>
                    <h3>{errorState.toUpperCase()}</h3>
                </ErrorMessage>
            )}
            </StyledLogin>
            
        </Layout>
    );
}
const StyledLogin = styled.div`
    display: flex;
    margin: auto;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    gap: 15px;
    height: 100%;
    width: fit-content;

    .loginPage-title {
        color: var(--text-100);
    }
`;
