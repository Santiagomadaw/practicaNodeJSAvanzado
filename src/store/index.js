import { applyMiddleware, combineReducers, createStore } from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';
import * as reducers from './reducers';
import * as actionCreators from './actions';
import { withExtraArgument } from 'redux-thunk';
import getAds from '../pages/AdvertsPage/service';
import { findHighestPrice } from '../pages/AdvertsPage/utils';
import { login, logout } from '../pages/login/services';
import { postAd } from '../pages/NewAdvertPage/services';
import { deleteAd, getAd } from '../pages/AdvertPage/service';
import getTags from '../components/layout/services';

const reducer = combineReducers(reducers);

// eslint-disable-next-line no-unused-vars
const failureRedirects = (router, pathMap) => store =>next => action =>{
  
        const result = next(action);
        if (!action.error) {
          return result;
        }
        const redirect = pathMap[action.payload.status];
        if (redirect) {
          router.navigate(redirect);
        }
        return result;
      };
    
const composeEnhancers = composeWithDevTools({ actionCreators });
const services={ getAds, login, logout, postAd, deleteAd, getAd, getTags }
export default function configureStore(preloadedState, { router }) {
  const store = createStore(
    reducer,
    preloadedState,
    composeEnhancers(
      applyMiddleware(
        withExtraArgument({ services,findHighestPrice, router }),
        failureRedirects(router, {
          401: '/login',
          404: '/404',
        }),
      ),
    ),
  );

  return store;
}