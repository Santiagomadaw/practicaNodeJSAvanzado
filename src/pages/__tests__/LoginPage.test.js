import '@testing-library/jest-dom';
import { act, render, screen } from '@testing-library/react';
import LoginPage from '../login/LoginPage';
import { Provider } from 'react-redux';
import { adCreate, authLogin } from '../../store/actions';
import userEvent from '@testing-library/user-event';
import NewAdvertPage from '../NewAdvertPage/NewAdvertPage';
import { BrowserRouter } from 'react-router-dom';
import { defaultState } from '../../store/reducers';

jest.mock('../../store/actions');

const userType = (input, text) => userEvent.type(input, text);
const userSelect = (input, options) => userEvent.selectOptions(input, options);
const userCheck = (input) => userEvent.click(input);


describe('LoginPage', () => {
  const state = { ui: { pending: false, error: null } };
  const store = {
    dispatch: () => { },
    getState: () => state,
    subscribe: () => { },
  };

  const renderComponent = () =>
    render(
      <Provider store={store}>
        <LoginPage />
      </Provider>,
    );

  test('Snap shot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('should dispatch authLogin action ', () => {
    const email = 'email';
    const password = 'password';
    renderComponent();

    const usernameInput = screen.getByRole('email');
    const passwordInput = screen.getByRole('password');
    const submitButton = screen.getByRole('button');
    expect(submitButton).toBeDisabled();

    act(() => userType(usernameInput, email));
    act(() => userType(passwordInput, password));
    expect(submitButton).toBeEnabled();
    userEvent.click(submitButton);
    expect(authLogin).toHaveBeenCalledWith({ email, password, save: false });


  });
});


describe('NewAdvertPage', () => {
  const state = {
    ...defaultState, tags: {
      loaded: true,
      data: [
        'lifestyle',
        'mobile',
        'motor',
        'work'
      ]
    }
  };
  const store = {
    dispatch: jest.fn(),
    getState: () => state,
    subscribe: () => { },
    location: location,
    navigate: jest.fn(),
    getTagsLoaded: () => state.tags,
  };

  const renderComponent = () =>
    render(
      <Provider store={store}>
        <BrowserRouter>
          <NewAdvertPage />
        </BrowserRouter>
      </Provider>,
    );


  test('Snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('should dispatch adCreate action', async () => {
    const name = 'name';
    const sale = true;
    const price = '100';
    const tags = 'work';

    renderComponent();

    const nameInput = screen.getByRole('name');
    const saleInput = screen.getByRole('checkbox');
    const priceInput = screen.getByRole('price');
    const tagsInput = screen.getByRole('tags');
    const submitButton = screen.getByRole('buttonSubmit');

    expect(submitButton).toBeDisabled();

    act(() => userType(nameInput, name));
    act(() => userCheck(saleInput, sale));
    act(() => userType(priceInput, price));
    act(() => userSelect(tagsInput, tags));

    expect(submitButton).toBeEnabled();

    userEvent.click(submitButton);

    expect(adCreate).toHaveBeenCalledWith({ name, sale, price, tags });
  });
});