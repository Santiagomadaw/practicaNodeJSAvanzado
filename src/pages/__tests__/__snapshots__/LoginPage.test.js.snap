// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage Snap shot 1`] = `
<div>
  <div
    class="sc-gsFSjX hZyESR loginPage"
  >
    <h1
      class="loginPage-title"
    >
      Login
    </h1>
    <form
      class="sc-aYaIB erRhpt"
      id="login-form"
    >
      <input
        class="sc-gEvDqW cZdiSF"
        id="email"
        name="email"
        role="email"
        type="text"
      />
      <input
        class="sc-gEvDqW cZdiSF"
        id="password"
        name="password"
        role="password"
        type="password"
      />
      <div
        class="sc-eqUzNf dxwZcj switch-container"
      >
        <h3
          class="switchtext"
        />
        <label
          class="switch"
        >
          <input
            id="buysellcheck"
            name="save"
            type="checkbox"
          />
          <span
            class="slider round"
          />
        </label>
        <h3
          class="switchtext"
        >
          Guardar
        </h3>
      </div>
      <button
        class="sc-dcJtft ggeaVg loginButton"
        disabled=""
        role="button"
        type="submit"
      >
        Login
      </button>
    </form>
  </div>
</div>
`;

exports[`NewAdvertPage Snapshot 1`] = `
<div>
  <section
    class="new-ad-contianer"
  >
    <form
      class="sc-aYaIB dTKQHr"
      id="ad-form"
    >
      <label
        for="name"
      >
        <h4>
          Articulo:
        </h4>
      </label>
      <input
        class="sc-gEvDqW lfZGwp"
        id="name"
        name="name"
        required=""
        role="name"
        type="text"
      />
      <label
        for="price"
      >
        <h4>
          Price:
        </h4>
      </label>
      <input
        class="sc-gEvDqW lfZGwp"
        id="price"
        name="price"
        required=""
        role="price"
        step="0.01"
        type="number"
      />
      <label
        for="tags"
      >
        <h4>
          Etiquetas:
        </h4>
      </label>
      <select
        class="sc-kAycRU JvKbt"
        id="tags"
        multiple=""
        name="tags"
        required=""
        role="tags"
      >
        <option
          id="lifestyle"
          value="lifestyle"
        >
          lifestyle
        </option>
        <option
          id="mobile"
          value="mobile"
        >
          mobile
        </option>
        <option
          id="motor"
          value="motor"
        >
          motor
        </option>
        <option
          id="work"
          value="work"
        >
          work
        </option>
      </select>
      <label
        for="photo"
      >
        <h4>
          Foto:
        </h4>
      </label>
      <input
        accept="image/png, image/jpeg"
        id="photo"
        name="photo"
        type="file"
      />
      <div
        class="sc-eqUzNf dxwZcj switch-container"
      >
        <h3
          class="switchtext"
        >
          Compra
        </h3>
        <label
          class="switch"
        >
          <input
            id="buysellcheck"
            name="sale"
            type="checkbox"
          />
          <span
            class="slider round"
          />
        </label>
        <h3
          class="switchtext"
        >
          Venta
        </h3>
      </div>
      <div
        class="buttonWrapper"
      >
        <button
          class="sc-dcJtft cwfUWM backButton"
          role="buttonBack"
          type="button"
        >
          Volver
        </button>
        <button
          class="sc-dcJtft cwfUWM newAdButton"
          disabled=""
          role="buttonSubmit"
          type="submit"
        >
          Crear anuncio
        </button>
      </div>
    </form>
  </section>
</div>
`;
